### PR TITLE
charm.URL: introduce channels.

### PR DESCRIPTION
**Pull request by frankban**
Include the channel in charm.URL.
For the time being, the only possible channel is "development".

### CODE REVIEW

File: url.go [hunk 1]
```diff
@@ -37,21 +37,37 @@ type URL struct {
 	Name     string // "wordpress"
 	Revision int    // -1 if unset, N otherwise
 	Series   string // "precise" or "" if unset
```
Comment (rogpeppe) [inline url.go]: Might be worth clarifying this actually:

    // "precise" or "" if unset; "bundle" if it's a bundle.

File: url.go [hunk 2]
```diff
@@ -37,21 +37,37 @@ type URL struct {
...
+	Channel  string // "development"
```
Comment (rogpeppe) [inline url.go]: // "development" or "" if no channel.

File: url.go [hunk 3]
```diff
@@ -37,21 +37,37 @@ type URL struct {
...
+const (
+	DevelopmentChannel = "development"
+)
+
 var (
-	validSeries = regexp.MustCompile("^[a-z]+([a-z0-9]+)?$")
-	validName   = regexp.MustCompile("^[a-z][a-z0-9]*(-[a-z0-9]*[a-z][a-z0-9]*)*$")
+	validSeries   = regexp.MustCompile("^[a-z]+([a-z0-9]+)?$")
+	validName     = regexp.MustCompile("^[a-z][a-z0-9]*(-[a-z0-9]*[a-z][a-z0-9]*)*$")
+	validChannels = map[string]bool{
```
Comment (rogpeppe) [inline url.go]: Since there's only one, why not just inline channel == DevelopmentChannel
until there's need for the map?

File: url.go [hunk 4]
```diff
@@ -37,21 +37,37 @@ type URL struct {
...
 	Revision int    // -1 if unset, N otherwise
 	Series   string // "precise" or "" if unset
+	Channel  string // "development"
 }
 
 var ErrUnresolvedUrl error = fmt.Errorf("charm or bundle url series is not resolved")
 
+// Channels represent different stages in the development of a charm or bundle.
+const (
+	DevelopmentChannel = "development"
```
Comment (mhilton) [inline url.go]: PublishedChannel = "" ?

File: url.go [hunk 5]
```diff
@@ -37,21 +37,37 @@ type URL struct {
 	Name     string // "wordpress"
 	Revision int    // -1 if unset, N otherwise
 	Series   string // "precise" or "" if unset
+	Channel  string // "development"
 }
 
 var ErrUnresolvedUrl error = fmt.Errorf("charm or bundle url series is not resolved")
 
+// Channels represent different stages in the development of a charm or bundle.
```
Comment (rogpeppe) [inline url.go]: perhaps:

type Channel string

const (
    DevelopmentChannel Channel = "development"
)

?

File: url.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ type Location interface {
 	String() string
 }
 
-// URL represents a charm location:
+// URL represents a charm or bundle location:
 //
 //     cs:~joe/oneiric/wordpress
 //     cs:oneiric/wordpress-42
```
Comment (rogpeppe) [inline url.go]: Let's add a channel example here too.

**Comment by mhilton** [comment]
:+1:

File: url.go [hunk 7]
```diff
@@ -143,9 +174,21 @@ func parseV1URL(url *gourl.URL, originalURL string) (*URL, error) {
 		}
 		r.User, parts = parts[0][1:], parts[1:]
 	}
+
+	// <channel>
+	if len(parts) > 1 {
+		if IsValidChannel(parts[0]) {
+			if r.Schema == "local" {
+				return nil, fmt.Errorf("local charm or bundle URL with channel: %q", originalURL)
```
Comment (rogpeppe) [inline url.go]: This doesn't read much like an error.

local charm or bundle URL %q is not allowed to have a channel

?

File: url.go (hunk 4 above)
Comment (rogpeppe) [inline url.go]: I think it's reasonable to leave the empty value hardcoded.

**Comment by rogpeppe** [comment]
LGTM with a few minor suggestions, thanks!

File: url.go (hunk 1 above)
Comment (frankban) [inline url.go]: Done.

File: url.go (hunk 2 above)
Comment (frankban) [inline url.go]: Done.

File: url.go (hunk 4 above)
Comment (frankban) [inline url.go]: +1

File: url.go (hunk 3 above)
Comment (urosj) [inline url.go]: +1

Comment (frankban) [inline url.go]: Ok, done.

File: url.go (hunk 6 above)
Comment (frankban) [inline url.go]: Done.

File: url.go (hunk 7 above)
Comment (frankban) [inline url.go]: This reflects the "error" returned for user and local charms.

File: url.go (hunk 5 above)
Comment (frankban) [inline url.go]: Ok I'll make this change,

**Comment by frankban** [comment]
Thanks for the reviews!
$$merge$$

File: url.go [hunk 8]
```diff
@@ -30,28 +30,47 @@ type Location interface {
...
-	User     string // "joe"
-	Name     string // "wordpress"
-	Revision int    // -1 if unset, N otherwise
-	Series   string // "precise" or "" if unset
+	Schema   string  // "cs" or "local".
+	User     string  // "joe".
+	Name     string  // "wordpress".
+	Revision int     // -1 if unset, N otherwise.
+	Series   string  // "precise" or "" if unset; "bundle" if it's a bundle.
+	Channel  Channel // "development" or "" if no channel.
```
Comment (urosj) [inline url.go]: :)

**Pull request merged**